### PR TITLE
Reload backgrounds and not just images

### PR DESCRIPTION
**Pull request by robin-carlier**
This adds the ability to reload backgrounds (and not just images). This was asked for in #290.

The code is mostly copypasta of the code for the image widget, with suitable modifications.

Since backgrounds are usually big, I also added an option to cross-fade between the old and the new background to avoid abrupt changes. The cross-fade can be customized with a float option `crossfade_time` for the widget. A negative value means no cross-fade.
The "mixing texture" shader I had to add for this can probably be used to implement similar cross-fades for image widgets, but I didn’t do it.

All timers/reloading options are ignored for screenshot-type backgrounds (perhaps that can be improved, but I’m not sure on how to handle reloading if `reload_cmd` is not supplied and the background is a screenshot).

I’m not a C++ dev nor a graphics programmer so pardon me if I did things the wrong way.

### CODE REVIEW

File: src/renderer/widgets/Background.hpp [hunk 1]
```diff
@@ -3,36 +3,68 @@
...
+    float                                   brightness        = 0.8172;
+    float                                   vibrancy          = 0.1696;
+    float                                   vibrancy_darkness = 0.0;
+    Vector2D                                viewport;
+    std::string                             path = "";
+
+    std::string                             resourceID;
+    std::string                             pendingResourceID;
+
+    float                                   crossfade_time = -1.0;
```
Comment (PaideiaDilemma) [inline src/renderer/widgets/Background.hpp:54]: camelCase

File: src/renderer/widgets/Background.cpp [hunk 2]
```diff
@@ -97,5 +171,131 @@ bool CBackground::draw(const SRenderData& data) {
...
\ No newline at end of file
+    if (fade)
+        return true; // actively redraw during fading
+    else
+        return data.opacity < 1.0;
+}
+
+void CBackground::plantTimer() {
+
+    RASSERT(!(reloadTime >= 0 && isScreenshot), "Reloadable background can't be used with screenshot")
```
[the remark this answers is not in the document]
Comment (robin-carlier) [reply src/renderer/widgets/Background.cpp:182]: Right, `plantTimer` is not supposed to be planted for screenshots anyways.

File: src/renderer/widgets/Background.hpp [hunk 3]
```diff
@@ -3,36 +3,68 @@
...
 class CBackground : public IWidget {
   public:
     CBackground(const Vector2D& viewport, COutput* output_, const std::string& resourceID, const std::unordered_map<std::string, std::any>& props, bool ss_);
+    ~CBackground();
 
     virtual bool draw(const SRenderData& data);
     void         renderRect(CColor color);
 
+    void         onTimerUpdate();
+    void         onFadeTimerUpdate();
```
[the remark this answers is not in the document]
Comment (PaideiaDilemma) [reply src/renderer/widgets/Background.hpp:33]: That comes straight from the image widget right? I guess you can leave it if that is the case.

[the remark this answers is not in the document]
Comment (robin-carlier) [reply src/renderer/widgets/Background.hpp]: The names come from the image widget yes, but this widget has also only one (reload) timer and doesn’t have the cross-fade stuff, so you’re right that it’d make more sense to name it this way. Plus I already changed it anyway.

File: src/renderer/widgets/Background.hpp [hunk 4]
```diff
@@ -3,36 +3,68 @@
...
+    CColor                                  color;
+    SPreloadedAsset*                        asset        = nullptr;
+    COutput*                                output       = nullptr;
+    bool                                    isScreenshot = false;
+    SPreloadedAsset*                        pendingAsset = nullptr;
+    bool                                    firstRender  = true;
+
+    std::unique_ptr<SFade>                  fade;
+
+    int                                     reloadTime;
```
Comment (vaxerski) [inline src/renderer/widgets/Background.hpp:65]: uninitialized primitive

File: src/renderer/widgets/Background.cpp [hunk 5]
```diff
@@ -97,5 +171,124 @@ bool CBackground::draw(const SRenderData& data) {
...
+        } else if (resourceID != pendingResourceID) {
+
+            pendingAsset = newAsset;
+
+            if (crossFadeTime > 0) {
+                // Start a fade
+                if (!fade) {
+
+                    fade = std::make_unique<SFade>(std::chrono::system_clock::now(), 0, nullptr);
+
```
Comment (vaxerski) [inline src/renderer/widgets/Background.cpp:269]: what are these empty lines all around here

File: src/renderer/widgets/Background.cpp [hunk 6]
```diff
@@ -97,5 +171,124 @@ bool CBackground::draw(const SRenderData& data) {
...
+        if (path.starts_with("file://"))
+            path = path.substr(7);
+
+        if (path.empty()) 
+            return;
+    }
+
+    try {
+        const auto MTIME = std::filesystem::last_write_time(path);
+        if (OLDPATH == path && MTIME == modificationTime) {
```
Comment (vaxerski) [inline src/renderer/widgets/Background.cpp:227]: no {} around short ifs

File: src/renderer/widgets/Background.cpp [hunk 7]
```diff
@@ -97,5 +171,124 @@ bool CBackground::draw(const SRenderData& data) {
...
+void CBackground::onCrossFadeTimerUpdate() {
+
+    // Animation done: Unload previous asset, deinitialize the fade and pass the asset
+
+    if (fade) {
+        fade->crossFadeTimer.reset();
+        fade.reset(nullptr);
+    };
+
+    if (!(blurPasses > 0 || isScreenshot)) {
```
Comment (vaxerski) [inline src/renderer/widgets/Background.cpp:194]: .

File: src/renderer/widgets/Background.cpp [hunk 8]
```diff
@@ -97,5 +171,124 @@ bool CBackground::draw(const SRenderData& data) {
...
+};
+
+void CBackground::onCrossFadeTimerUpdate() {
+
+    // Animation done: Unload previous asset, deinitialize the fade and pass the asset
+
+    if (fade) {
+        fade->crossFadeTimer.reset();
+        fade.reset(nullptr);
+    };
```
Comment (vaxerski) [inline src/renderer/widgets/Background.cpp:192]: ; ?

Comment (robin-carlier) [reply src/renderer/widgets/Background.cpp:192]: Nixlang has apperently corrupted my mind and now I put those everywhere. Sorry I’ll fix them up.

File: src/renderer/widgets/Background.cpp (hunk 6 above)
Comment (robin-carlier) [reply src/renderer/widgets/Background.cpp:227]: Do I also remove them if they are followed by a long `else` block?

**Review by vaxerski — APPROVED** [review]
thanks! Wiki mr needed

**Review by vaxerski — APPROVED** [review]

**Pull request merged**